Fix NTLM auth for requests with KeepAlive == false

kartofeell opened the pull request:
NTLM Authentication requires connection to be maintained after successful
authentication. If client sets KeepAlive to false on an instance
of HttpWebRequest, when connecting to a server with NTLM authentication,
'Connection: close' header should NOT be included in the second step of handshake.

Before the fix, the fix, HttpWebRequest would keep
'Connection: Close' header, which would cause authentication failure.

More details in following article:
http://www.benjaminathawes.com/2010/10/14/ntlms-dependency-on-http-keep-alives-another-cause-of-the-dreaded-401-1-error/

monoadmin commented:
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

kartofeell commented:
Opened a new bug for this issue: https://bugzilla.xamarin.com/show_bug.cgi?id=27214

knocte commented on mcs/class/System/System.Net/HttpWebRequest.cs:
```diff
@@ -1172,6 +1172,12 @@ string GetHeaders ()
+                        // http://www.benjaminathawes.com/2010/10/14/ntlms-dependency-on-http-keep-alives-another-cause-of-the-dreaded-401-1-error/
+                        if (auth_state.NtlmAuthState == NtlmAuthState.Challenge && version == HttpVersion.Version11) {
+                                webHeaders.RemoveAndAdd (connectionHeader, "keep-alive");
+                        }
```
Use tabs, not spaces.

kartofeell replied:
Yup, apologies, fixing.

akoeplinger commented:
build

migueldeicaza commented:
Thanks for your contribution!

directhex commented:
This introduces a test failure:

https://jenkins.mono-project.com/job/test-mono-pull-request/494/label=debian-amd64/testReport/junit/MonoTests.System.Net/HttpWebRequestTest/NtlmAuthentication_KeepAlive/

Note that it's entirely possible that the test is wrong - but please take a look

akoeplinger commented:
I think it's just a flaky test as the i686 build doesn't show it. It should be investigated nonetheless.

kartofeell commented:
At the moment it is passing on my local build (amd64 on Ubuntu 14.04). I'll have a closer look.

kartofeell commented:
The fact that tests were passing locally on my machine and on one of the machines on CI was a coincidence. I started to see failures when I switched my machine to use only one CPU core, and when I updated tests to simulate the real scenario more accurately, I noticed that the fix was not working.

It seems that I missed one place which required a change. I had to make some minor updates in `WebConnection` class, because it was closing a socket connection during the authentication when KeepAlive was set to false

I have updated the commit with better fix and updated tests. Please have a look.

akoeplinger commented:
approve

kartofeell commented:
I had a quick look at the regressions reported by CI (before the results were deleted), and it seems to me that the failures are not related to changes I made. 

Two failing tests reported as a regression do not cover the code which was changed in my commit:
 - `System.Net.HttpListenerBugs.TestNonChunkedAsync`
 - `System.Threading.ManualResetEventSlimTests.Constructor_Defaults`

I do not know what may be causing these failures :/

alexrp commented:
build

lewurm commented:
is there still interest to merge this or should it be closed?

monojenkins commented:
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

kartofeell commented:
I think this PR should be closed. There are some tests still failing on ARM, and I did not have chance to debug and fix the problem.
Please correct me if I am wrong. As far, I know, there is some work going on that focuses on porting parts of network stack from .NET to Mono, which will most likely result in this issue being fixed.

lewurm commented:
build

lewurm commented:
build

lewurm commented:
I'm closing this PR, feel free to reopen this PR if you have some time to work on it. Thank you!
